java/arcs/core/data/expression/EvaluatorParticle.kt

```diff
@@ -0,0 +1,120 @@
+        metadata.particleName,
+        schemaMap.mapValues {
+            (name, schema) ->
+            setOf(object : EntitySpec<Entity> {
+                override val SCHEMA = schema
+
+                override fun deserialize(data: arcs.core.data.RawEntity) =
+                    EntityBase(SCHEMA.name.toString(), SCHEMA).apply {
+                        // TODO: Take nested entities into account.
+                        deserialize(data)
+                    }
+            })
+        }
+    )
+
+    override fun onReady() {
+        val scope = scopeFromReadHandles()
+        evaluateWriteHandleExpressions(scope)
+    }
+
+    private fun scopeFromReadHandles(): CurrentScope<Any> {
+        val scopeMap = mutableMapOf<String, Any>()
+
+        metadata.handles.filterValues { hc -> hc.mode.canRead }.forEach { (name, hc) ->
+            val handle = handles.handles[name]
+            val schema = requireNotNull(schemaMap[name])
+            when (hc.type) {
+                is SingletonType<*> -> {
+                    val singletonHandle = handle as ReadSingletonHandle<EntityBase>
+                    val entity = singletonHandle.fetch()
+                    if (entity != null) {
+                        scopeMap[name] = entityAsScope(entity, schema)
+                    }
+                }
+                is CollectionType<*> -> {
+                    val collectionHandle = handle as ReadCollectionHandle<EntityBase>
+                    val entities = collectionHandle.fetchAll()
+                    scopeMap[name] = entities.map { entityAsScope(it, schema) }
+                }
+            }
+        }
+
+        return CurrentScope(scopeMap)
+    }
+
+    private fun evaluateWriteHandleExpressions(handlesScope: CurrentScope<Any>) {
+        metadata.handles.filterValues { hc -> hc.mode.canWrite }.forEach { (name, hc) ->
+            val handle = handles.handles[name]
+            val schema = requireNotNull(schemaMap[name])
+            val expression = requireNotNull(metadata.handles[name]).expression!!
+            val result = evalExpression(expression, handlesScope)
+            when (hc.type) {
+                is SingletonType<*> -> {
+                    val singletonHandle = handle as WriteSingletonHandle<EntityBase>
+                    singletonHandle.store(entityFromScope(result as MapScope<*>, schema))
+                }
+                is CollectionType<*> -> {
+                    val collectionHandle = handle as WriteCollectionHandle<EntityBase>
+                    (result as Sequence<MapScope<*>>).forEach {
+                        collectionHandle.store(entityFromScope(it, schema))
+                    }
+                }
+            }
+        }
+    }
+
+    private fun entityAsScope(entity: EntityBase, schema: Schema): MapScope<*> =
+        // TODO: Support collection fields.
+        schema.fields.singletons.mapValues { (name, _) ->
+            entity.getSingletonValue(name)
+        }.asScope()
+
```

Comment:
Eventually we will replace this by supplying a ScopeBuilder to the Evaluator that can set the fields directly without an intermediate map. Not important now, but will be when we have a lot of data and iterating over largish collections

Reply:
Added TODO